### PR TITLE
[CI] Improve coverage cache

### PR DESCRIPTION
**Pull request by amenasria**
### What does this PR do?

This PR improves the cache coverage granularity. The way the cache used to work on a dev branch:
- Go through all modules:
    - If a `coverage.out` file exists on the dev branch, leave it as is.
    - Else add the `coverage.out` file from main branch.
- Upload the `coverage.out` files batch to Codecov.

This was imprecise for huge modules for example for `/` the dev branch file would be really small but huge for the main branches. That's why I added more granularity, we're now rewriting lines in the `coverage.out` file and not just the whole file.

The new approach is:
- Rename the dev branch `coverage.out` files to `coverage.out.dev`.
- Go through all modules:
    - If a `coverage.out` file exists on the main branch replace all lines related to the modules in the `coverage.out.dev` files.
    - Else rename `coverage.out.dev` to `coverage.out`.
- Upload the `coverage.out` files batch to Codecov.

### Motivation

The initial motivation was to get a more accurate cache because we had a lot of false positive indirect changes.

We also had another problem with the coverage:

On each commit on developer branches, some non-modified files are reporting a modified coverage. Looking at the files, the issue seems to impact unit tests testing code with non deterministic behavior, for example:
- Time-dependent code: depending in which order the routines are executed you’ll get different behaviors (e.g: function creating an object if it doesn’t exist but not doing anything if it does exist).
- Code using randomness without using a seed: the behavior will be different at each run, so the logic might enter different blocks during the execution
- Code interacting with external systems: a unit test that pings Datadog’s backend will have a different behavior during/outside an incident.
- Code using syscalls: getting the datetime, read the execution time of a process…

This non-deterministic code execution path will generate flakey coverage (let's name them "coverage flakes").

A nice side effect about the new cache mechanism is that such flakey behaviour will happen only when developers are modifying modules with non-deterministic code. This is not optimal as they might have modified unrelated code so it shouldn't raise on their PR but that the least worst case IMO. Plus it might allow us to raise awareness on those badly designed unit tests.
 

### Additional Notes

https://datadoghq.atlassian.net/browse/ACIX-289
https://datadoghq.atlassian.net/browse/ACIX-314

### CODE REVIEW

File: tasks/coverage.py [hunk 1]
```diff
@@ -176,12 +177,51 @@ def upload_coverage_to_s3(ctx: Context):
...
+def _get_dev_coverage_files(dev_cov_lines: str) -> set[str]:
+    browsed_dev_files = set()
+    for line in dev_cov_lines:
+        file_path = line.split(':')[0]
+        browsed_dev_files.add(file_path)
+    return browsed_dev_files
+
+
+def _merge_dev_in_main_coverage(main_cov_file: str, dev_cov_file: str) -> None:
+    with open(main_cov_file, encoding='utf-8') as main_cov:
```
Comment (CelianR) [inline tasks/coverage.py:189]: **🥜 nitpick:** ‏I'm not sure the encoding arg is useful, is it?

**Review by CelianR — APPROVED** [review]

File: tasks/coverage.py (hunk 1 above)
Comment (amenasria) [reply tasks/coverage.py:189]: The linter complains if I don't use it :') https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/unspecified-encoding.html

File: tasks/coverage.py [hunk 2]
```diff
@@ -176,12 +177,51 @@ def upload_coverage_to_s3(ctx: Context):
...
-def apply_missing_coverage(ctx: Context, from_commit_sha: str, debug: bool = False):
+def _get_dev_coverage_files(dev_cov_lines: str) -> set[str]:
+    browsed_dev_files = set()
+    for line in dev_cov_lines:
+        file_path = line.split(':')[0]
+        browsed_dev_files.add(file_path)
+    return browsed_dev_files
+
+
+def _merge_dev_in_main_coverage(main_cov_file: str, dev_cov_file: str) -> None:
```
Comment (pducolin) [inline tasks/coverage.py:188]: **💬 suggestion**
Can you add a tiny example of what these files look like?

**Review by pducolin — APPROVED** [review]

File: tasks/coverage.py [hunk 3]
```diff
@@ -121,7 +122,7 @@ def upload_to_codecov(
 
     if pull_coverage_cache:
         with gitlab_section("Applying missing coverage cache from S3", collapsed=True):
-            apply_missing_coverage(ctx, from_commit_sha=get_main_parent_commit(ctx), debug=debug)
+            apply_missing_coverage(ctx, from_commit_sha=get_main_parent_commit(ctx), keep_temp_files=debug)
```
[the remark this answers is not in the document]
Comment (amenasria) [reply tasks/coverage.py]: Good point ! Fixed in [07bc3ea](https://github.com/DataDog/datadog-agent/pull/27407/commits/07bc3ea705980874cb6ee9a63b8802a472c766e8): I made it `--debug-cache` to be more explicit.

**Pull request merged**